pkg/kubectl/cmd/set/set_port.go

```diff
@@ -0,0 +1,281 @@
+	o.protocol = strings.ToUpper(cmdutil.GetFlagString(cmd, "protocol"))
+	o.containerSelector = cmdutil.GetFlagString(cmd, "containers")
+	o.overwrite = cmdutil.GetFlagBool(cmd, "overwrite")
+	o.Cmd = cmd
+
+	cmdNamespace, enforceNamespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+
+	builder := f.NewBuilder(!o.Local).
+		ContinueOnError().
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(enforceNamespace, &o.FilenameOptions).
+		Flatten()
+	if !o.Local {
+		builder = builder.
+			SelectorParam(o.Selector).
+			ResourceTypeOrNameArgs(o.All, args...).
+			Latest()
+	}
+	o.Infos, err = builder.Do().Infos()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *PortOptions) Validate() error {
+	errors := []error{}
+	if o.protocol != "TCP" && o.protocol != "UDP" {
+		errors = append(errors, fmt.Errorf("protocol must tcp or udp"))
+	}
+	if o.port < 0 {
+		errors = append(errors, fmt.Errorf("port must greater than 0"))
+	}
+	return utilerrors.NewAggregate(errors)
+}
+
+func (o *PortOptions) Run() error {
+	allErrs := []error{}
+
+	patches := CalculatePatches(o.Infos, o.Encoder, func(info *resource.Info) ([]byte, error) {
+		transformed := false
+		_, err := o.UpdatePodSpecForObject(info.Object, func(spec *api.PodSpec) error {
+			containers, _ := selectContainers(spec.Containers, o.containerSelector)
+			if len(containers) == 0 {
+				fmt.Println(info.Mapping.Resource, info.Name, o.containerSelector)
+				_, err := fmt.Fprintf(o.Err, "warning: %s/%s does not have any containers matching %q\n", info.Mapping.Resource, info.Name, o.containerSelector)
+				if err != nil {
+					return err
+				}
+				return nil
+			}
+
+			ports, err := o.portsForObject(info.Object)
+			if err != nil {
+				return err
+			}
+			for _, c := range containers {
+				if !o.overwrite {
+					for _, port := range ports {
+						if port == strconv.Itoa(o.port) {
+							return fmt.Errorf("'%s' already has a port %v, and --overwrite is false", c.Name, port)
+						}
+					}
+				}
+				for _, port := range c.Ports {
+					if port.ContainerPort != int32(o.port) || string(port.Protocol) != o.protocol {
+						transformed = true
+					}
+				}
+				containerPort := api.ContainerPort{
+					ContainerPort: int32(o.port),
+					Protocol:      api.Protocol(o.protocol),
+				}
+				c.Ports = append(c.Ports, containerPort)
+			}
+			if err != nil {
+				return err
+			}
+			return nil
+		})
+		if !transformed {
+			if _, err := fmt.Fprintln(o.Out, "no resources changed"); err != nil {
+				return nil, err
+			}
+			return nil, err
+		}
+		if transformed && err == nil {
+			return runtime.Encode(o.Encoder, info.Object)
+		}
+		return nil, err
+	})
+
+	for _, patch := range patches {
+		info := patch.Info
+		if patch.Err != nil {
+			allErrs = append(allErrs, fmt.Errorf("error: %s/%s %v\n", info.Mapping.Resource, info.Name, patch.Err))
+			continue
+		}
+
+		// no changes
+		if string(patch.Patch) == "{}" || len(patch.Patch) == 0 {
```

Comment:
should have a better way of `string(patch.Patch) == "{}"`.

Reply:
this follow other set command